nestkernel/connection_manager.cpp

```diff
@@ -1221,23 +1221,23 @@ nest::ConnectionManager::get_connections(
       ConnectorBase* connections = connections_[ tid ][ syn_id ];
       if ( connections != NULL )
       {
+        const size_t num_connections_in_thread = connections->size();
+        for ( index lcid = 0; lcid < num_connections_in_thread; ++lcid )
+        {
+          const index source_gid = source_table_.get_gid( tid, syn_id, lcid );
+          connections->get_connection_with_specified_targets( source_gid,
+            target_neuron_gids,
+            tid,
+            lcid,
+            synapse_label,
+            conns_in_thread );
+        }
+
```

Comment:
as far as I can tell this loop uses at most `n_conns * n_targets` comparisons (loop + linear complexity of `std::find` in `get_connection_with_specified_targets`). the previous implementation also should use the same amount of comparisons in the worst case. your new implementation is much faster, so where did my reasoning fail?

Reply:
In the previous implementation both the vector of specified targets and the vector of local connections are always completely iterated, so there are always `n_conns * n_targets` comparisons, like you say. This implementation iterates over the connections first, so it only has to check if the target of the connection is among the specified targets, and therefore doesn't have to check all of them for every connection. Also `std::find` may do some hardware-near optimizations.

Reply:
ok, I think I understand. while the worst case runtime is identical for both implementations, the average case is much better in the new one. :+1: